### PR TITLE
[CDAP-8666] add version in run records, workflow node state and token when upgrade

### PR DESCRIPTION
**Pull request by ecfm**
https://issues.cask.co/browse/CDAP-8666
http://builds.cask.co/browse/CDAP-RUT692-2

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 1]
```diff
@@ -680,43 +680,98 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
+
+    // If typeOfT is RunRecordMeta.class and originalKey starts with
+    // recordType:namespace:appName:appVersion:programType, no need to upgrade
+    if (typeOfT.equals(RunRecordMeta.class) && keyParts.size() > 4) {
+      try {
+        ProgramType.valueOf(keyParts.get(4));
+        // no need to upgrade if no IllegalArgumentException is thrown, return originalKey directly
+        return originalKey;
+      } catch (IllegalArgumentException e) {
+        // expected when the version is not part of the originalKey
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: It would be good to log this that its an expected exception

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 2]
```diff
@@ -680,43 +680,98 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
-      // max parts = 4 => appMeta.ns.app.version
-      for (int i = 0; i < 4; i++) {
-        splitter.skipString();
-        keyParts++;
+      // try to iterate through all key parts of type String
+      while (true) {
+        keyParts.add(splitter.getString());
+      }
+    } catch (BufferUnderflowException e) {
+      // expected when the originalKey is iterated to the end or non-String type key part is encountered
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: Add log statement here.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 1 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: added in next line

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 2 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: added in next line

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 3]
```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
   }
 
   public void upgradeVersionKeys() {
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: It would be good to not do all of these in same transaction. Can you use batching to update the keys?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 4]
```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
-    Map<MDSKey, ApplicationMeta> appMetas = listKV(startKey, ApplicationMeta.class);
-    Map<MDSKey, ApplicationMeta> newAppMetas = new HashMap<>();
-    for (Map.Entry<MDSKey, ApplicationMeta> oldKey : appMetas.entrySet()) {
-      MDSKey newKey = appendDefaultVersion(oldKey.getKey());
-      newAppMetas.put(newKey, oldKey.getValue());
+    upgradeVersionKeys(TYPE_APP_META, ApplicationMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_STARTED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_SUSPENDED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_COMPLETED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_NODE_STATE, RunRecordMeta.class);
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: Are we using version info in query? Can you verify?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 5]
```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
-  private static MDSKey appendDefaultVersion(MDSKey oldKey) {
-    MDSKey.Splitter splitter = oldKey.split();
-    int keyParts = 0;
+  // Append version only if it doesn't have version in the key
+  private static MDSKey appendDefaultVersion(String recordType, Type typeOfT, MDSKey originalKey) {
+    List<String> keyParts = new ArrayList<>();
+    MDSKey.Splitter splitter = originalKey.split();
     try {
-      // max parts = 4 => appMeta.ns.app.version
-      for (int i = 0; i < 4; i++) {
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: Why did you remove this and using while?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 6]
```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
-        splitter.skipString();
-        keyParts++;
+      // try to iterate through all key parts of type String
+      while (true) {
+        keyParts.add(splitter.getString());
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: Is it guaranteed that `splitter.getString()` will throw exception? What happens if it does not?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 4 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: Yes, because all the queries in `AppMetadataStore` calls `getProgramKeyBuilder` to construct MDSKey, which always includes app version

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java [hunk 7]
```diff
@@ -266,6 +266,30 @@ public void run(DatasetContext context) throws Exception {
...
+   * @param transactional the {@link Transactional} to use for transactional execution.
+   * @param timeoutInSeconds transaction timeout in seconds
+   * @param callable the {@link TxCallable} to be executed inside a transaction
+   * @param <V> type of the result
+   * @return value returned by the given {@link TxCallable}.
+   * @throws TransactionFailureException if failed to execute the given {@link TxRunnable} in a transaction
+   *
+   * TODO: CDAP-6103 Move this to {@link Transactional} when revamping tx supports in program.
+   */
+  public static <V> V execute(Transactional transactional, int timeoutInSeconds,
```
Comment (sagarkapare) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java]: where is this method getting called?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java [hunk 8]
```diff
@@ -446,6 +470,26 @@ private static TransactionControl getTransactionControl(Class<?> cls, String met
...
+   * likely indicate poor exception handling.
+   *
+   * @param transactional the {@link Transactional} used to submit the task
+   * @param timeoutInSeconds transaction timeout in seconds
+   * @param callable the task
+   * @param <V> type of result
+   * @return value returned by the given {@link TxCallable}.
+   * @throws RuntimeException for any errors that occur
+   */
+  public static <V> V executeUnchecked(Transactional transactional, int timeoutInSeconds,
```
Comment (sagarkapare) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/Transactions.java]: is this method required?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java [hunk 9]
```diff
@@ -863,7 +863,8 @@ void clear() throws Exception {
   }
 
   public void upgradeAppVersion() throws Exception {
-    Transactions.executeUnchecked(transactional, new TxRunnable() {
+    Transactions.executeUnchecked(transactional, Integer.getInteger(configuration.get(Constants.UPGRADE_TIMEOUT_SECS)),
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java]: can use `configuration.getInt(Constants.UPGRADE_TIMEOUT_SECS)`

File: cdap-common/src/main/resources/cdap-default.xml [hunk 10]
```diff
@@ -419,6 +419,14 @@
     </description>
   </property>
 
+  <property>
+    <name>upgrade.timeout.secs</name>
```
Comment (sagarkapare) [inline cdap-common/src/main/resources/cdap-default.xml]: better call `app.meta.upgrade.timeout.sec`? since its not the upgrade timeout but part of it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: javadoc for the public method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 11]
```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
-      newAppMetas.put(newKey, oldKey.getValue());
+    upgradeVersionKeys(TYPE_APP_META, ApplicationMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_STARTED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_SUSPENDED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_COMPLETED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_NODE_STATE, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_TOKEN, RunRecordMeta.class);
+  }
+
+  public <T> void upgradeVersionKeys(String recordType, Type typeOfT) {
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: javadocs for the public method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 12]
```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
+      write(newEntry.getKey(), newEntry.getValue());
     }
   }
 
-  // Append version only if it doesn't have version at the end
-  private static MDSKey appendDefaultVersion(MDSKey oldKey) {
-    MDSKey.Splitter splitter = oldKey.split();
-    int keyParts = 0;
+  // Append version only if it doesn't have version in the key
+  private static MDSKey appendDefaultVersion(String recordType, Type typeOfT, MDSKey originalKey) {
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: I feel this method needs more documentation as to key format before version change and key format after that.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 13]
```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
+    upgradeVersionKeys(TYPE_APP_META, ApplicationMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_STARTED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_SUSPENDED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_RUN_RECORD_COMPLETED, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_NODE_STATE, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_TOKEN, RunRecordMeta.class);
+  }
+
+  public <T> void upgradeVersionKeys(String recordType, Type typeOfT) {
+    MDSKey startKey = new MDSKey.Builder().add(recordType).build();
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: I think its better to have switch case on recordType and have separate methods for appending versions for different record types

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 14]
```diff
@@ -680,43 +680,100 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
+    Map<MDSKey, T> oldMap = listKV(startKey, typeOfT);
+    Map<MDSKey, T> newMap = new HashMap<>();
+    for (Map.Entry<MDSKey, T> oldEntry : oldMap.entrySet()) {
+      MDSKey newKey = appendDefaultVersion(recordType, typeOfT, oldEntry.getKey());
+      newMap.put(newKey, oldEntry.getValue());
     }
 
     //Delete old rows
     deleteAll(startKey);
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:710]: if we delete all keys and something goes wrong after this we will not be able to recover.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 6 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: better do 
```java
while (true) {
 try {
  keyParts.add(splitter.getString());
  } catch (BufferUnderflowException e) {
 // expected when the originalKey is iterated to the end or non-String type key part is encountered
  break;
  }
}

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 2 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: No need to log when expected right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 15]
```diff
@@ -52,6 +52,7 @@
 import com.google.common.reflect.TypeToken;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
+import org.apache.hadoop.mapreduce.v2.app.webapp.App;
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: why need this?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 16]
```diff
@@ -679,44 +680,158 @@ public boolean apply(RunRecordMeta runRecordMeta) {
...
+    upgradeVersionKeys(TYPE_WORKFLOW_NODE_STATE, RunRecordMeta.class);
+    upgradeVersionKeys(TYPE_WORKFLOW_TOKEN, RunRecordMeta.class);
+  }
+
+  /**
+   * Upgrades the rowkeys for the given record type.
+   *
+   * @param recordType type of the record
+   * @param typeOfT    content type of the record
+   * @param <T>        content type of the record
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: nit: type param

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java (hunk 14 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:710]: should nt affect since its happening in transaction

File: cdap-common/src/main/resources/cdap-default.xml [hunk 17]
```diff
@@ -419,6 +419,14 @@
     </description>
   </property>
 
+  <property>
+    <name>app.meta.upgrade.timeout.secs</name>
```
Comment (sagarkapare) [inline cdap-common/src/main/resources/cdap-default.xml:423]: this change requires updating DEFAULT_XML_MD5_HASH  in `admin-manual/build.sh`

**Comment by sagarkapare** [comment]
code LGTM. one comment about the build hash. please run the doc build as well.
Merge once tested.
